packages/web3/src/block/block.ts

```diff
@@ -20,117 +20,153 @@ import { Subscription, SubscribeOptions } from '../utils/subscription'
+    const blocks: node.BlockEntry[] = []
+    let lastBlock = toBlock
+    while (lastBlock.height - 1 > fromHeight) {
+      const parentHash = this.getParentHash(lastBlock)
+      const block = await this.getBlockByHash(parentHash)
+      blocks.push(block)
+      lastBlock = block
+    }
+    const parentHash = this.getParentHash(lastBlock)
+    if (parentHash !== fromHash) {
+      await this.handleReorg(toBlock.chainFrom, toBlock.chainTo, fromHash, parentHash)
+    }
+    return blocks.reverse()
+  }
+
+  private async handleBlocks(blocks: node.BlockEntry[][]) {
+    const allBlocks: node.BlockEntry[] = []
+    for (let index = 0; index < blocks.length; index += 1) {
+      const blocksPerChain = blocks[index]
+      if (blocksPerChain.length === 0) continue
+
+      allBlocks.push(...blocksPerChain)
+      const parent = this.parents[index]
+      if (parent !== undefined) {
+        const missingBlocks = await this.getMissingBlocksAndHandleReorg(parent.hash, parent.height, blocksPerChain[0])
+        allBlocks.push(...missingBlocks)
+      }
+      const latestBlock = blocksPerChain[blocksPerChain.length - 1]
+      this.parents[index] = { hash: latestBlock.hash, height: latestBlock.height }
+    }
+
+    const sortedBlocks = allBlocks.sort((a, b) => a.timestamp - b.timestamp)
+    await this.messageCallback(sortedBlocks)
+  }
+
   override async polling(): Promise<void> {
     try {
-      const chainInfo = await this.nodeProvider.blockflow.getBlockflowChainInfo({
-        fromGroup: this.fromGroup,
-        toGroup: this.toGroup
-      })
-
-      while (this.currentBlockHeight <= chainInfo.currentHeight) {
-        const hashes = await this.getHashesAtHeight(this.currentBlockHeight)
-        const block = await this.getBlockByHash(hashes[0])
-        if (this.parentBlockHash !== undefined && this.getParentHash(block) !== this.parentBlockHash) {
-          await this.handleReorg(this.parentBlockHash, this.currentBlockHeight - 1)
+      const now = Date.now()
+      if (this.fromTimeStamp >= now) return
+
+      while (this.fromTimeStamp < now) {
+        if (this.isCancelled()) return
+        const toTs = Math.min(this.fromTimeStamp + DEFAULT_INTERVAL, now)
+        const result = await this.nodeProvider.blockflow.getBlockflowBlocks({ fromTs: this.fromTimeStamp, toTs })
+        const isEmpty = result.blocks.every((blocksPerChain) => blocksPerChain.length === 0)
+        if (!isEmpty) {
+          this.fromTimeStamp = toTs + 1
+          await this.handleBlocks(result.blocks)
+          continue
         }
-        await this.messageCallback(block)
-        this.currentBlockHeight += 1
-        this.parentBlockHash = hashes[0]
+        // there are no blocks within the time range [fromTimeStamp, fromTimeStamp + DEFAULT_INTERVAL]
+        // update the `fromTimeStamp` to avoid infinite loop
+        if (toTs !== now) this.fromTimeStamp += DEFAULT_INTERVAL
```

Comment:
I initially did the same, but I encountered some issues during testing. For example, if the current time is `now` and the polling interval is 1 second, there is a block with a timestamp between `[now - 1s, now]`. But the full node might save this block after `now`, resulting in being unable to subscribe to the block in time.

Reply:
Even so, it may still cause some block subscriptions to be delayed. We're going to get these blocks here: https://github.com/alephium/alephium-web3/pull/367/files#diff-1b61443c0aeb7c5d14698e1e63680ba37608d4718af5958ae3ba2fcd00074f0cR115. I'm not sure if there's a better way to handle it.

Reply:
Another approach is to poll each chain based on its height, but this would require more network requests.